Sources/SwiftKafka/KafkaConsumer.swift

```diff
@@ -151,27 +160,34 @@ public final class KafkaConsumer {
         }
     }
 
+    deinit {
+        // This occurs e.g. when for await loop is exited through break
+        self.shutdownGracefully()
+    }
+
     /// Subscribe to the given list of `topics`.
     /// The partition assignment happens automatically using `KafkaConsumer`'s consumer group.
     /// - Parameter topics: An array of topic names to subscribe to.
     /// - Throws: A ``KafkaError`` if subscribing to the topic list failed.
     private func subscribe(topics: [String]) throws {
-        assert(!self.closed)
-
-        for topic in topics {
-            rd_kafka_topic_partition_list_add(
-                self.subscribedTopicsPointer,
-                topic,
-                KafkaPartition.unassigned.rawValue
-            )
-        }
+        let action = self.stateMachine.withLockedValue { $0.setUpConnection() }
+        switch action {
+        case .setUpConnection(let client, let subscribedTopicsPointer):
+            for topic in topics {
+                rd_kafka_topic_partition_list_add(
+                    subscribedTopicsPointer,
+                    topic,
+                    KafkaPartition.unassigned.rawValue
+                )
+            }
 
-        let result = self.client.withKafkaHandlePointer { handle in
-            rd_kafka_subscribe(handle, self.subscribedTopicsPointer)
-        }
+            let result = client.withKafkaHandlePointer { handle in
+                rd_kafka_subscribe(handle, subscribedTopicsPointer)
+            }
```

Comment:
For all these places where we use `withKafkaHandlePointer` can we just add a method on the `KafkaClient` instead? At best that method also already does the conversion to the correct error.